macros/src/main/scala/gestalt/macros/JsonMacros.scala

```diff
@@ -0,0 +1,74 @@
+import scala.gestalt._
+
+object JsonMacros {
+
+  sealed trait JsValue
+  final case class JsString(value: String) extends JsValue
+
+  final case class JsObject(items: Seq[(String, JsValue)]) extends JsValue {
+    def firstValue(key: String) = items.find(_._1 == key).map(_._2)
+  }
+
+  trait Format[A] {
+    def toJson(o: A): JsValue
+    def fromJson(json: JsValue): Option[A]
+  }
+
+  /**
+    * Creates a {{{Format}}} for the case class.
+    * Limitations:
+    * 1. does not handle cyclic references, i.e. {{{case class A(field: A, ...)}}}
+    * 2. does not handle optional arguments
+    * @tparam T type of the case class
+    * @return new anonymous Format class
+    */
+  def format[T](): Format[T] = meta {
+    import toolbox._
+    val tpe: Type = T.tpe
+    if (!tpe.isCaseClass) {
+      error("Not a case class", T.pos)
+      q"???"
+    } else {
+      val fields = tpe.caseFields
+      val namesAndTypes = fields.map {
+        f => f.name -> f.info
+      }
+
+      case class JsonItem(name: String, pairOut: TermTree, readOption: ValDef, implicitFormat: Option[ValDef])
+      val jsonItems: Seq[JsonItem] = namesAndTypes.map {
+        case (name, stringType) if stringType =:= Type.typeRef("java.lang.String") =>
+          JsonItem(name,
+            pairOut = Tuple(Seq(Lit(name), q"JsString(${Select(Ident("o"), name)})")),
+            readOption = q"val $name = obj.firstValue(${Lit(name)}).collect{case JsString(value) => value}",
+            implicitFormat = None
+          )
+        case (name, otherType) =>
+          val implFormaterName = name + "_formatter"
+          JsonItem(name,
+            pairOut = Tuple(Seq(Lit(name), q"${Ident(implFormaterName)}.toJson(${Select(Ident("o"), name)})")),
+            readOption = q"val $name = obj.firstValue(${Lit(name)}).flatMap(x =>${Ident(implFormaterName)}.fromJson(x))",
+            implicitFormat = Some(q"val $implFormaterName=implicitly[Format[${otherType.toTree}]]")
+          )
+      }
+      val allDefined = q"${jsonItems.map(i => q"${Ident(i.name)}.isDefined").reduceLeft((a, b) => q"$a && $b")}"
+      val construction = NewInstance(T, Seq(jsonItems.map(i => q"${Ident(i.name)}.get")))
```

Comment:
@valdisxp1 A potential problem here is portability -- the expression `new p.C[T](args)` is very special, different compiler implementations can easily differ on how the expression is represented. In your case, it seems easy to extract relevant parts from `T`?

As this extractor is potentially useful in other places, we can create a helper extractor `PathType` to do the job. WDYT?